godot-core/src/builtin/array.rs

```diff
@@ -185,61 +282,189 @@ impl Array {
         let end = end.min(len);
         let step = step.unwrap_or(1);
         assert!(step != 0);
-        self.as_inner()
-            .slice(to_i64(begin), to_i64(end), step.try_into().unwrap(), true)
+        let slice: Array =
+            self.as_inner()
+                .slice(to_i64(begin), to_i64(end), step.try_into().unwrap(), true);
+        // SAFETY: slice() returns a typed array with the same type as Self
+        unsafe { slice.assume_type() }
+    }
+
+    /// Appends another array at the end of this array. Equivalent of `append_array` in GDScript.
+    pub fn extend_array(&mut self, other: TypedArray<T>) {
+        // SAFETY: Read-only arrays are covariant: conversion to a variant array is fine as long as
+        // we don't insert values into it afterwards, and `append_array()` doesn't do that.
```

Comment:
And most importantly, `T` is the same type as `Self` has 🙂 (maybe mention that)

I was first confused, but then realized that the `assume_type()` only does an "upcast" to `TypedArray<Variant>`.

Reply:
I added an explicit `::<Variant>` to make that clearer.

Whether or not `T` is the same type as `Self` does not matter for the safety or unsafety of the upcast.

Reply:
> I added an explicit `::<Variant>` to make that clearer.

Good idea!

> Whether or not `T` is the same type as `Self` does not matter for the safety or unsafety of the upcast.

True, I was thinking of the `self.as_inner().append_array(other)` rather than the upcast -- but even that is safe, as Godot runtime-checks the types. So you're right, we can leave the description.